Error recovery for case

jez opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->


### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

Sorry, I didn't go through the work of making this one into small commits.


### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

froydnj commented on parser/parser/cc/grammars/typedruby.ypp:
```diff
@@ -1865,15 +1866,40 @@ lbrace_cmd_block_start:
 			                  else_ ? else_->nod : nullptr, $4);
+                      driver.rewind_if_dedented($1, $4);
+                    }
+                | kCASE expr_value opt_terms kEND
```
Can we format this so that we have whitespace where `case_body` would go, so it's more obvious how this alternative differs from existing code?

```suggestion
                | kCASE expr_value opt_terms           kEND
```

froydnj commented on parser/parser/cc/grammars/typedruby.ypp:
```diff
@@ -1865,15 +1866,40 @@ lbrace_cmd_block_start:
+                      // even if we aren't in indentationAware mode.
+                      driver.rewind_if_dedented($1, $4, /*force*/true);
+                    }
+                | kCASE opt_terms kEND
```
Same request here for the formatting.

```suggestion
                | kCASE            opt_terms           kEND

```

froydnj commented on parser/parser/cc/grammars/typedruby.ypp:
```diff
@@ -1865,15 +1866,40 @@ lbrace_cmd_block_start:
                         &case_body->whens,
                         else_ ? else_->tok : nullptr,
 			                  else_ ? else_->nod : nullptr, $5);
+                      driver.rewind_if_dedented($1, $5);
```
It's not obvious to me why we need to alter the behavior of the pattern-matching `case` here, since this was existing non-error code.  Shouldn't this be removed?

froydnj commented on parser/parser/cc/grammars/typedruby.ypp:
```diff
@@ -1865,15 +1866,40 @@ lbrace_cmd_block_start:
+                      $$ = driver.build.case_error(self, $1, cond, nullptr);
+                      driver.replace_last_diagnostic(dlevel::ERROR, dclass::UnexpectedToken, $1, driver.token_name($1->type()));
                     }
-                 | kCASE expr_value opt_terms p_case_body kEND
```
This change seems unnecessary.

jez replied:
It was the only production in this file that didn’t have its pipe character aligned so I fixed it.

froydnj commented on parser/Builder.cc:
```diff
@@ -614,10 +614,27 @@ class Builder::Impl {
 
     unique_ptr<Node> case_(const token *case_, unique_ptr<Node> expr, sorbet::parser::NodeVec whenBodies,
                            const token *elseTok, unique_ptr<Node> elseBody, const token *end) {
+        auto loc = tokLoc(case_);
+        if (end != nullptr) {
+            loc = loc.join(tokLoc(end));
+        }
         return make_unique<Case>(tokLoc(case_).join(tokLoc(end)), std::move(expr), std::move(whenBodies),
```
Surely this `loc` addition would be better if it was used in the `make_unique` call?  Or was this an experiment that turned out to be unnecessary?

(Slightly concerned that tests didn't seem to exercise the `end == nullptr` case, too, but maybe that is just a consequence of this being dead code.)

jez replied:
Yeah, I started trying to reuse the `case_` builder for the error handling cases, and then at some point realized that it wouldn't be needed. This code is dead.

froydnj commented on parser/parser/cc/grammars/typedruby.ypp:
```diff
@@ -1856,6 +1856,7 @@ lbrace_cmd_block_start:
                         &case_body->whens,
                         else_ ? else_->tok : nullptr,
 			                  else_ ? else_->nod : nullptr, $5);
+                      driver.rewind_if_dedented($1, $5);
```
Why are we rewinding in existing productions?

froydnj commented on parser/parser/cc/grammars/typedruby.ypp:
```diff
@@ -1865,15 +1866,40 @@ lbrace_cmd_block_start:
                         &case_body->whens,
                         else_ ? else_->tok : nullptr,
 			                  else_ ? else_->nod : nullptr, $4);
+                      driver.rewind_if_dedented($1, $4);
```
Why are we rewinding in existing productions?

froydnj reviewed (changes requested):
I think this is reasonable, but I have concerns about:

1. `rewind_if_dedented` now appearing in rules that aren't actually errors, which seems like scope creep.
2. various things that don't seem to be used or changes that appear spurious.

Will stamp once those concerns are cleared up.

jez commented:
I've fixed the small changes. Thanks for pointing those out.

To answer all the "Why are we rewinding in existing productions?" questions:

That is the whole point of the change. Calling `rewind_if_dedented` in those cases:

- will only do anything if `driver.indentationAware` is set, which is only set when we're doing a second parse to attempt better error recovery
- is crucial to make every test that I added in this PR pass—none of them pass without this change. Every production rule that I've modified in this PR has at least one, but usually many more, tests.

jez commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_L7diw9SiuA6mWC
→ https://go/builds/bui_L7di9yer8NQ7mG
